### PR TITLE
File importer

### PR DESCRIPTION
**Pull request by wochinge**
**Proposed changes**:
- fixes https://github.com/RasaHQ/rasa/issues/3937
- move `SkillSelector` to `rasa/data/skill` (I messed up the git rename, I'm sorry!)
- updated `SkillSelector` to use `TrainingFileImporter` interface
- new interface `TrainingFileImporter` which centralizes the file importing
- made nlu training asynchronously
- updated training data validator to use TrainingFileImporter

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

**Comment by wochinge** [comment]
@tabergma Would be great to get some first feedback from you! 
There are a couple of more things we could base on the new interface, but maybe it makes sense to tackle this in another issue (interactive learning, some testing parts).

**Comment by wochinge** [comment]
I will fix the remaining tests tomorrow

File: rasa/importers/skill.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
+import logging
```
Comment (wochinge) [inline rasa/importers/skill.py]: I messed up the renaming here, sorry! I removed all the static stuff and adapted it to the `TrainingFileImporter` interface.

**Comment by tabergma** [comment]
This PR is huge. Can we somehow split it up? Maybe just introduce the new importers in the first PR and move the actual usage of them into another PR?

File: rasa/importers/importer.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        return importer
+
+    @staticmethod
+    def _importer_from_dict(
+        importer_config: Dict,
+        config_path: Text,
+        domain_path: Optional[Text] = None,
+        training_data_paths: Optional[List[Text]] = None,
+    ) -> Optional["TrainingFileImporter"]:
+        from rasa.importers.skill import SkillSelector
```
Comment (tabergma) [inline rasa/importers/importer.py]: Why is the `SkillSelector` in a different file if it is also an importer?

File: rasa/importers/importer.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        else:
+            try:
+                importer_class = common_utils.class_from_module_path(module_path)
+            except (AttributeError, ImportError):
+                logging.warning("Importer '{}' not found.".format(module_path))
+                return None
+
+        import rasa.cli.utils as cli_utils
+
+        constructor_arguments = cli_utils.minimal_kwargs(
```
Comment (tabergma) [inline rasa/importers/importer.py]: I guess, we should move this method (`minimal_kwargs`) to `utils.common` as it is not only used in the cli package. But should maybe happen in a different PR.

File: rasa/importers/importer.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+            self.config = {}
+
+        self._domain_path = domain_path
+
+        if training_data_paths:
+            self._story_files, self._nlu_files = data.get_core_nlu_files(
+                training_data_paths
+            )
+
+    async def get_config(self, **kwargs: Optional[Dict[Text, Any]]) -> Dict:
```
Comment (tabergma) [inline rasa/importers/importer.py]: Why did you add `**kwargs: Optional[Dict[Text, Any]]` as argument?

File: rasa/importers/utils.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from typing import Iterable, Text
+
+from rasa.nlu.training_data import TrainingData
+
+
+def training_data_from_paths(paths: Iterable[Text], language: Text) -> TrainingData:
+    from rasa.nlu.training_data import loading
+
+    training_datas = [loading.load_data(nlu_file, language) for nlu_file in paths]
```
Comment (tabergma) [inline rasa/importers/utils.py]: `load_data` can throw a `ValueError` if the file does not exist. Can this happen here? Should we add a try/except`

File: rasa/importers/skill.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+from rasa.importers.importer import TrainingFileImporter
+from rasa.importers import utils
+from rasa.nlu.training_data import TrainingData
+from rasa.core.training.structures import StoryGraph
+
+logger = logging.getLogger(__name__)
+
+
+class SkillSelector(TrainingFileImporter):
+    def __init__(
```
Comment (tabergma) [inline rasa/importers/skill.py]: Just for my understanding: All the parameters need to be set in the config file?

File: rasa/cli/interactive.py [hunk 7]
```diff
@@ -107,8 +107,8 @@ def check_training_data(args):
         get_validated_path(f, "data", DEFAULT_DATA_PATH, none_is_valid=True)
         for f in args.data
     ]
-    story_directory, nlu_data_directory = data.get_core_nlu_directories(training_files)
-    if not os.listdir(story_directory) or not os.listdir(nlu_data_directory):
+    story_files, nlu_files = data.get_core_nlu_files(training_files)
```
Comment (tabergma) [inline rasa/cli/interactive.py]: Why was this changed?

File: rasa/core/validator.py [hunk 8]
```diff
@@ -22,17 +23,15 @@ def __init__(self, domain: Domain, intents: TrainingData, stories: List[StorySte
...
     @classmethod
-    async def from_files(
-        cls, domain_file: Text, nlu_data: Text, story_data: Text
-    ) -> "Validator":
+    async def from_importer(cls, importer: TrainingFileImporter) -> "Validator":
         """Create an instance from the domain, nlu and story files."""
 
-        domain = Domain.load(domain_file)
+        domain = await importer.get_domain()
         asyncio.new_event_loop()
```
Comment (tabergma) [inline rasa/core/validator.py]: Not part of the PR: But why is that line here?

**Comment by wochinge** [comment]
> This PR is huge. Can we somehow split it up? Maybe just introduce the new importers in the first PR and move the actual usage of them into another PR?

@tabergma Sorry, I know it's painful 😞  But just adding the importers would not make them usable at all. Only putting them in the train process makes them usable at all 🤔

File: rasa/cli/interactive.py (hunk 7 above)
Comment (wochinge) [reply rasa/cli/interactive.py]: we don't need to move this in a tmp directory, we can just check whether any files were found.

File: rasa/importers/importer.py (hunk 2 above)
Comment (wochinge) [reply rasa/importers/importer.py]: I wanted to avoid to having one large module. Would you suggest a different split?

File: rasa/importers/importer.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+
+
+class TrainingFileImporter:
+    """Common interface for different mechanisms to load training data."""
+
+    async def get_domain(self) -> Domain:
+        """Retrieves the domain which should be used for the training."""
+        raise NotImplementedError()
+
+    async def get_story_data(
```
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/importers/importer.py]: I decided to use a `StoryGraph` which holds the `stories`. I renamed it cause I thought stories isn't really accurate then. But I can rename it 👍

File: rasa/importers/importer.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+    async def get_nlu_data(self, language: Optional[Text] = "en") -> TrainingData:
+        """Retrieves the nlu training data which should be used for the training."""
+        raise NotImplementedError()
+
+    @staticmethod
+    def load_from_config(
+        config_path: Text,
+        domain_path: Optional[Text] = None,
+        training_data_paths: Optional[List[Text]] = None,
+        load_only_nlu_data: bool = False,
```
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/importers/importer.py]: I had a lot of warning logs when doing `train nlu`. I'm not having anything similar for `train core`, so I think we don't need that atm.

File: rasa/importers/importer.py (hunk 4 above)
Comment (wochinge) [reply rasa/importers/importer.py]: legacy, removed it.

File: rasa/importers/importer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/importers/importer.py]: But it might help to speed up training if you just run `rasa train core`, wouldn't it? We would not need to load the NLU data. But, as no NLU data are provided (or they are filtered out), it would not change anything. So, I guess, you are right, we don't need it 😄

File: rasa/importers/importer.py (hunk 2 above)
Comment (tabergma) [reply rasa/importers/importer.py]: Yeah, make sense. But maybe we should then move also the `SimpleFileImporter` to a separate file as it the only "real" importer next to the `SkillSelector`.

File: rasa/importers/importer.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+    async def get_nlu_data(self, language: Optional[Text] = "en") -> TrainingData:
+        return utils.training_data_from_paths(self._nlu_files, language)
+
+    async def get_domain(self) -> Domain:
+        domain = Domain.empty()
+        try:
+            domain = Domain.load(self._domain_path)
+            domain.check_missing_templates()
+        except Exception:
```
Comment (tabergma) [inline rasa/importers/importer.py]: Why `Exception` and not `InvalidDomain`?

File: rasa/importers/skill.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        if os.path.isfile(path):
+            self._init_from_file(path)
+        elif os.path.isdir(path):
+            self._init_from_directory(path)
+
+    def _init_from_file(self, path: Text) -> None:
+        from rasa import data  # pytype: disable=pyi-error
+
+        path = os.path.abspath(path)
+        if os.path.exists(path) and data.is_config_file(path):
```
Comment (tabergma) [inline rasa/importers/skill.py]: Should we log at least a warning if the file was not found?

File: rasa/importers/skill.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+        for parent, _, files in os.walk(path):
+            for file in files:
+                full_path = os.path.join(parent, file)
+                if not self.is_imported(full_path):
+                    # Check next file
+                    continue
+
+                if data.is_domain_file(full_path):
+                    self._domain_paths.append(full_path)
+                elif data._is_nlu_file(full_path):
```
Comment (tabergma) [inline rasa/importers/skill.py]: Should we make those files "public"? (`_is_nlu_file` and `_is_story_file`)

File: rasa/model.py [hunk 14]
```diff
@@ -192,24 +184,24 @@ def model_fingerprint(
...
         FINGERPRINT_CONFIG_CORE_KEY: _get_hash_of_config(
-            config_file, include_keys=CONFIG_MANDATORY_KEYS_CORE
+            config, include_keys=CONFIG_MANDATORY_KEYS_CORE
         ),
         FINGERPRINT_CONFIG_NLU_KEY: _get_hash_of_config(
-            config_file, include_keys=CONFIG_MANDATORY_KEYS_NLU
+            config, include_keys=CONFIG_MANDATORY_KEYS_NLU
         ),
-        FINGERPRINT_DOMAIN_KEY: domain_hash,
-        FINGERPRINT_NLU_DATA_KEY: _get_hashes_for_paths(nlu_data),
```
Comment (tabergma) [inline rasa/model.py]: `_get_hashes_for_paths` is not used anymore and can be removed

File: rasa/nlu/config.py [hunk 15]
```diff
@@ -18,24 +18,31 @@ def __init__(self, message: Text) -> None:
...
+
+    if config is not None:
         try:
-            file_config = rasa.utils.io.read_config_file(filename)
+            file_config = rasa.utils.io.read_config_file(config)
         except yaml.parser.ParserError as e:
             raise InvalidConfigError(
-                "Failed to read configuration file "
-                "'{}'. Error: {}".format(filename, e)
+                "Failed to read configuration file " "'{}'. Error: {}".format(config, e)
```
Comment (tabergma) [inline rasa/nlu/config.py]:
```suggestion
                "Failed to read configuration file '{}'. Error: {}".format(config, e)
```

File: rasa/train.py [hunk 16]
```diff
@@ -149,37 +127,30 @@ def handle_domain_if_not_exists(
...
         Path of the trained model archive.
     """
-    new_fingerprint = model.model_fingerprint(
-        config, domain, nlu_data_directory, story_directory
-    )
+    new_fingerprint = await model.model_fingerprint(file_importer)
 
-    dialogue_data_not_present = not os.listdir(story_directory)
-    nlu_data_not_present = not os.listdir(nlu_data_directory)
+    dialogue_data = await file_importer.get_story_data()
```
Comment (tabergma) [inline rasa/train.py]: Would also rename `dialogue_data` to `stories` here. Also in the print messages.

**Comment by tabergma** [comment]
@wochinge I added a couple of more comments. Still need to test it locally. Will do it tomorrow. But in general, I like the idea and the implementation looks good. 
There is only one thing I'm not sure about (also added a comment): Where are the constructor arguments of the importers set? Currently, it looks like they need to be specified in the config file itself.

File: rasa/importers/importer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/importers/importer.py]: mhm, the nlu data would still be loaded for the fingerprint. So maybe makes sense to add a core only importer 👍

File: rasa/importers/importer.py (hunk 2 above)
Comment (wochinge) [reply rasa/importers/importer.py]: 👍

File: rasa/importers/utils.py (hunk 5 above)
Comment (wochinge) [reply rasa/importers/utils.py]: the file has to exist, cause we collected it earlier. There would have to be a delete right between finding and loading it. I don't think we'd have to check that extra.

File: rasa/core/validator.py (hunk 8 above)
Comment (wochinge) [reply rasa/core/validator.py]: threw it out

File: rasa/importers/skill.py (hunk 6 above)
Comment (wochinge) [reply rasa/importers/skill.py]: it tries to pass in `config_path`, `domain_path`, `training_data_paths` which are given through the cli, the rest has to be provided through the config file.

File: rasa/importers/importer.py (hunk 11 above)
Comment (wochinge) [reply rasa/importers/importer.py]: 👍 had to change it in `Domain` so that it also gives an `InvalidDomain` in case it's not found

File: rasa/importers/importer.py [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+    """Common interface for different mechanisms to load training data."""
+
+    async def get_domain(self) -> Domain:
+        """Retrieves the domain which should be used for the training."""
+        raise NotImplementedError()
+
+    async def get_story_data(
+        self,
+        interpreter: "NaturalLanguageInterpreter" = RegexInterpreter(),
+        template_variables: Optional[Dict] = None,
```
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/importers/importer.py]: @akelad Do you have idea? I did some digging in the code and apparently you can specify some templates in stories which are replaced while reading the files?

**Comment by wochinge** [comment]
@tabergma Ready for a final review
@erohmensing @MetcalfeTom Do you have some input on the docs part of the pr?

**Comment by erohmensing** [comment]
Content looks good, thanks for the full example! Will do a quick copy edit and fix the docs warnings while i'm at it (check those next time 😉)

File: docs/api/training-data-importing.rst [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+
+    class MyImporter(TrainingFileImporter):
+        """Example for a custom importer component."""
+
+        def __init__(
+                self,
+                config_file: Optional[Text] = None,
+                domain_path: Optional[Text] = None,
+                training_data_paths: Optional[Union[List[Text], Text]] = None,
+                **kwargs: Dict
```
Comment (erohmensing) [inline docs/api/training-data-importing.rst]: @wochinge `Dict[Text, Any]`? or just `Any`? (no need to update, just comment)

File: rasa/importers/importer.py [hunk 19]
```diff
@@ -21,7 +21,7 @@ class TrainingFileImporter:
         """Retrieves the domain of the bot.
 
         Returns:
-            Loaded `Domain`.
+            Loaded ``Domain``.
```
Comment (wochinge) [inline rasa/importers/importer.py]: @erohmensing I'm pretty sure that it's only single ticks in pydoc (that's not rich text)

File: rasa/importers/importer.py [hunk 20]
```diff
@@ -32,39 +32,39 @@ class TrainingFileImporter:
...
         Returns:
-            `StoryGraph` containing all loaded stories.
+            ``StoryGraph`` containing all loaded stories.
         """
 
         raise NotImplementedError()
 
-    async def get_config(self) -> Dict:
-        """"Retrieves the configuration which should be used for the training.
+    async def get_config(self) -> Dict[Text, Any]:
```
Comment (wochinge) [inline rasa/importers/importer.py]: @erohmensing This could literally be anything. So could also have some int keys.

**Comment by wochinge** [comment]
If anybody has input for better names for `TrainingFileImporter` and `SimpleFileImporter` I'd be very happy. Not really satisfied with the naming atm.

File: rasa/importers/importer.py (hunk 19 above)
Comment (erohmensing) [reply rasa/importers/importer.py]: the way it renders on the docs page needs the double tics -- here is double
Screenshot 2019-07-19 at 14 54 44
vs single 
Screenshot 2019-07-19 at 14 55 16
with `make livedocs`

File: rasa/importers/importer.py (hunk 20 above)
Comment (erohmensing) [reply rasa/importers/importer.py]: got it. sorry!

File: rasa/importers/importer.py (hunk 19 above)
Comment (wochinge) [reply rasa/importers/importer.py]: Ah ok 🤔 Thing the plugin should handle this instead of changing things in our code.

Comment (erohmensing) [reply rasa/importers/importer.py]: Can look into that in the docs sprint 😀

Comment (erohmensing) [reply rasa/importers/importer.py]: In general i'm pretty unhappy with how our autoclass and automethod look tbh

File: rasa/importers/skill.py (hunk 6 above)
Comment (tabergma) [reply rasa/importers/skill.py]: ok, I guess I missed that part, will have another look

**Comment by MetcalfeTom** [comment]
Docs be fine :+1:

File: rasa/importers/importer.py [hunk 21]
```diff
@@ -0,0 +1,274 @@
...
+        elif module_path == SkillSelector.__name__:
+            importer_class = SkillSelector
+        else:
+            try:
+                importer_class = common_utils.class_from_module_path(module_path)
+            except (AttributeError, ImportError):
+                logging.warning("Importer '{}' not found.".format(module_path))
+                return None
+
+        constructor_arguments = rasa.utils.common.minimal_kwargs(
```
Comment (tabergma) [inline rasa/importers/importer.py]:
```suggestion
        constructor_arguments = common_utils.minimal_kwargs(
```

File: rasa/importers/importer.py [hunk 22]
```diff
@@ -0,0 +1,274 @@
+import asyncio
+from functools import reduce
+from typing import Text, Optional, List, Dict
+import logging
+
+import rasa.utils.common
```
Comment (tabergma) [inline rasa/importers/importer.py]: can be removed as you have further down `import rasa.utils.common as common_utils`

File: rasa/importers/skill.py [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+        )
+
+    def _init_from_path(self, path: Text) -> None:
+        if os.path.isfile(path):
+            self._init_from_file(path)
+        elif os.path.isdir(path):
+            self._init_from_directory(path)
+
+    def _init_from_file(self, path: Text) -> None:
+        from rasa import data  # pytype: disable=pyi-error
```
Comment (tabergma) [inline rasa/importers/skill.py]: already imported

File: rasa/importers/skill.py [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+
+        # import config files from paths which have not been processed so far
+        for p in import_candidates:
+            self._init_from_path(p)
+
+    def _is_explicitly_imported(self, path: Text) -> bool:
+        return not self.no_skills_selected() and self.is_imported(path)
+
+    def _init_from_directory(self, path: Text):
+        from rasa import data  # pytype: disable=pyi-error
```
Comment (tabergma) [inline rasa/importers/skill.py]: already imported

File: tests/test_importer.py [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+        lambda: importer.get_stories(),
+        lambda: importer.get_nlu_data(),
+        lambda: importer.get_domain(),
+    ]
+    for f in functions_to_test:
+        with pytest.raises(NotImplementedError):
+            await f()
+
+
+async def test_simple_file_importer(project: Text):
```
Comment (tabergma) [inline tests/test_importer.py]:
```suggestion
async def test_rasa_file_importer(project: Text):
```

Comment (tabergma) [reply tests/test_importer.py]: same for more test further down

File: docs/api/training-data-importers.rst [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+                config_file: Path to configuration file from command line arguments.
+                domain_path: Path to domain file from command line arguments.
+                training_data_paths: Path to training files from command line arguments.
+                **kwargs: Extra parameters passed through configuration in configuration file.
+            """
+
+            pass
+
+        async def get_domain(self) -> Domain:
+            path_to_domain_file = self._custom_get_domain_file()
```
[the remark this answers is not in the document]
Comment (wochinge) [reply docs/api/training-data-importers.rst]: added them with `pass`, e.g.

```
def _custom_get_domain_file(self) -> Text:
            pass
```
